src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java

```diff
@@ -1282,34 +1061,190 @@ else if (fieldClass == AccessModifierOption[].class) {
+                    if (list.get(i)) {
+                        if (first) {
+                            first = false;
+                        }
+                        else {
+                            sb.append(", ");
+                        }
+
+                        sb.append(TokenUtil.getTokenName(i));
+                    }
+                }
+
+                result = sb.toString();
+            }
+            else {
+                final StringBuilder sb = new StringBuilder(20);
+
+                for (int i = 0; i < Array.getLength(value); i++) {
+                    if (first) {
+                        first = false;
+                    }
+                    else {
+                        sb.append(", ");
+                    }
+
+                    sb.append(TokenUtil.getTokenName((int) Array.get(value, i)));
+                }
+
+                result = sb.toString();
+            }
+        }
+        else {
+            result = Arrays.toString((int[]) value).replace("[", "").replace("]", "");
+        }
+        if (result.isEmpty()) {
+            result = "{}";
+        }
+        return result;
+    }
+
     /**
      * Checks if the given property is takes token names as a type.
      *
-     * @param sectionName The name of the section/module being worked on.
-     * @param propertyName The property name to work with.
+     * @param field The backed field of the section/module being worked on.
      * @return {@code true} if the property is takes token names as a type.
-     * @noinspection OverlyComplexBooleanExpression
      */
-    private static boolean isPropertyTokenType(String sectionName, String propertyName) {
-        return "AtclauseOrder".equals(sectionName) && "target".equals(propertyName)
-            || "IllegalType".equals(sectionName) && "memberModifiers".equals(propertyName)
-            || "MagicNumber".equals(sectionName)
-                    && "constantWaiverParentToken".equals(propertyName)
-            || "MultipleStringLiterals".equals(sectionName)
-                    && "ignoreOccurrenceContext".equals(propertyName)
-            || "DescendantToken".equals(sectionName) && "limitedTokens".equals(propertyName);
+    private static boolean isPropertyTokenType(Field field) {
+        final BeanPropertyType propertyType = field.getDeclaredAnnotation(BeanPropertyType.class);
+        return propertyType != null && propertyType.value() == PropertyType.TOKEN_ARRAY;
     }
 
-    private static Field getField(Class<?> clss, String propertyName) {
+    /**
+     * Returns the bean property's field.
+     *
+     * @param fieldClass The bean property's type
+     * @param propertyName The bean property's name
+     * @return the bean property's field
+     */
+    private static Field getField(Class<?> fieldClass, String propertyName) {
         Field result = null;
+        Class<?> currentClass = fieldClass;
 
-        if (clss != null) {
+        while (!Object.class.equals(currentClass)) {
```

Comment:
Refactored to avoid recursion and to avoid calls to `setAccessible` on `Object` class.